### ISSUE TITLE
Implement gas metering abstraction

### ISSUE THREAD

**Issue by seunlanlege** [ComposableFi/ibc-relayer#6]
So currently we naively relay all packets to the counter party chain. In production its more likely that these batches will be to expensive (probably even more than is allowed into a block) to relay all at once. So we probably want to first estimate the gas cost somehow  then split them up into batches based on this estimate.

**Comment by blasrodri** [comment]
This is an interesting space. Is there any resource we can look at in order to understand how this has been done by others?

**Comment by seunlanlege** [comment]
Thought about this a bit more and we can definitely implement this tx queue in the relayer core rather than for every chain.

We'll just introduce a new api to chain 

```rust

trait Chain {
    const BlockMaxExecutionWeight: u64;

    fn estimate_execution_weight(&self, msgs: Vec<Any>) -> u64;

...

**Comment by seunlanlege** [comment]
> This is an interesting space. Is there any resource we can look at in order to understand how this has been done by others?

I unfortunately don't have this sir, but I don't think that the implementation details of a gas aware tx queue are so complex they require a reference.

**Comment by blasrodri** [comment]
can we leverage some SDKs/RPC endpoints?
https://docs.near.org/concepts/basics/transactions/gas-advanced#accurate-estimates-with-automated-tests
https://docs.nethereum.com/en/latest/nethereum-estimating-gas/

**Comment by seunlanlege** [comment]
Yes this issue is based on the idea that most blockchains have an api for estimating gas costs.

**Comment by blasrodri** [comment]
Perfect. Thanks
